File: python/with_java/programmers/level 2/주차 요금 계산.py
def solution(fees, records):
    inInfo = {}
    timeInfo = {}
    for record in records:
        time, num, cmd = record.split()
        time = list(map(int, time.split(":")))
        time = time[0] * 60 + time[1]
        if cmd == "IN":
            inInfo[num] = time
        else:
            timeInfo[num] = timeInfo.get(num, 0) + time - inInfo[num]
            del inInfo[num]
    for num in inInfo:
        timeInfo[num] = timeInfo.get(num, 0) + 23 * 60 + 59 - inInfo[num]
    
    timeInfo = sorted(list(timeInfo.items()), key=lambda x:x[0])
    answer = [fees[1] for _ in range(len(timeInfo))]
    for i in range(len(timeInfo)):
        num, time = timeInfo[i]
        if time < fees[0]:
            continue
        time -= fees[0]
        per = 0
        if time % fees[2] != 0:
            per += 1
        per += time // fees[2]
        answer[i] += per * fees[3]
    
    return answer

File: java/src/main/java/ps_traning/programmers/level_2/주차_요금_계산.java
package ps_traning.programmers.level_2;

import java.util.*;

public class 주차_요금_계산 {
    public int[] solution(int[] fees, String[] records) {
        int[] answer = {};
        Map<String, Integer> inInfo = new HashMap<>();
        Map<String, Integer> timeInfo = new HashMap<>();
        for (String record : records) {
            String[] r = record.split(" ");
            int time = 0;
            String num = r[1];
            String cmd = r[2];
            String[] temp = r[0].split(":");
            time = Integer.valueOf(temp[0]) * 60 + Integer.valueOf(temp[1]);
            if (cmd.equals("IN")) inInfo.put(num, time);
            else {
                timeInfo.put(num, timeInfo.getOrDefault(num, 0) + time - inInfo.get(num));
                inInfo.remove(num);
            }
        }
        for (String num : inInfo.keySet()) {
            timeInfo.put(num, timeInfo.getOrDefault(num, 0) + 23 * 60 + 59 - inInfo.get(num));
        }
        List<String> sortedNum = new ArrayList<>(timeInfo.keySet());
        Collections.sort(sortedNum);
        answer = new int[timeInfo.size()];

        for (int i = 0; i < timeInfo.size(); i++) {
            answer[i] += fees[1];
            String num = sortedNum.get(i);
            int time = timeInfo.get(num);
            if (time < fees[0]) continue;
            time -= fees[0];
            int per = 0;
            if (time % fees[2] != 0) per++;
            per += (int) (time / fees[2]);
            answer[i] += per * fees[3];
        }

        return answer;
    }
}
